### PR TITLE
fix calcProfit and test

### PR DESCRIPTION
**Pull request by taiki1frsh**

### CODE REVIEW

**Comment by mkXultra** [comment]
@taiki1frsh 
use your implmemnt function to closePotision.

**Comment by taiki1frsh** [comment]
@mkXultra 
done.

File: x/derivatives/keeper/perpetual_futures.go [hunk 1]
```diff
@@ -77,54 +77,28 @@ func (k Keeper) ClosePerpetualFuturesPosition(ctx sdk.Context, position types.Po
...
-
-		if closedRate.LTE(openedRate) {
-			profit := positionInstance.Size_.Sub(closedRate.Mul(sdk.NewDec(int64(positionInstance.Leverage))))
-			profitAmount := profit.Quo(*closedRate)
-
-			amountToUser = principal.Add(profitAmount)
-		} else {
-			loss := closedRate.Mul(sdk.NewDec(int64(positionInstance.Leverage))).Sub(positionInstance.Size_)
-			lossAmount := loss.Quo(*closedRate)
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, position.Address.AccAddress(), types.ModuleName, sdk.Coins{sdk.NewCoin(position.Market.BaseDenom, feeAmount.RoundInt())}) // TODO: this is wrong.
```
Comment (mkXultra) [inline x/derivatives/keeper/perpetual_futures.go:85]: @taiki1frsh 
What happens if the owner of position has a zero balance?
Wouldn't it be better to take the commission from remaing margin?

File: x/derivatives/types/positions.go [hunk 2]
```diff
@@ -84,12 +85,9 @@ func (m PerpetualFuturesPosition) NeedLiquidation(minMarginMaintenanceRate, curr
 func (m PerpetualFuturesPosition) EffectiveMargin(currentBaseUsdRate, currentQuoteUsdRate, currentMarginUsdRate sdk.Dec) sdk.Dec {
 	effectiveMargin := sdk.NewDecFromInt(m.RemainingMargin.Amount).Mul(currentBaseUsdRate.Quo(currentMarginUsdRate))
```
Comment (mkXultra) [inline x/derivatives/types/positions.go:86]: effective margin fanc is exists in feat/run_msgreport_in_beginblock
plz use it

File: x/derivatives/keeper/perpetual_futures.go [hunk 3]
```diff
@@ -77,54 +77,28 @@ func (k Keeper) ClosePerpetualFuturesPosition(ctx sdk.Context, position types.Po
...
-		}
-		break
-	case types.PositionType_POSITION_UNKNOWN:
-		return fmt.Errorf("unknown position type")
+	futuresPosition := types.NewPerpetualFuturesPosition(position, positionInstance)
+	returningAmount := futuresPosition.CalcReturningAmountAtClose(*closedRate)
+	// If loss is over the margin, it means liquidity provider takes the loss.
+	if returningAmount.IsNegative() {
+		returningAmount = sdk.ZeroInt()
+		// emit event to tell how much loss is taken by liquidity provider.
```
Comment (mkXultra) [inline x/derivatives/keeper/perpetual_futures.go:95]: plz move this code to CalcReturningAmountAtClose.
Business logic should be confined within types.

**Review by mkXultra — CHANGES_REQUESTED** [review]

File: x/derivatives/keeper/perpetual_futures.go (hunk 3 above)
Comment (taiki1frsh) [reply x/derivatives/keeper/perpetual_futures.go:95]: then, what's the best way that we treat the amount of the loss to the liquidity provider?
add a returning value containing possible loss to liquidity provider?

File: x/derivatives/keeper/perpetual_futures.go (hunk 1 above)
Comment (taiki1frsh) [reply x/derivatives/keeper/perpetual_futures.go:85]: i didn't touch this.
i just added error case handling.
should i care this in this PR?

File: x/derivatives/types/positions.go (hunk 2 above)
Comment (taiki1frsh) [reply x/derivatives/types/positions.go:86]: where should I apply this fanc?
what is the original intention of this fanc?

File: x/derivatives/keeper/perpetual_futures.go (hunk 3 above)
Comment (mkXultra) [reply x/derivatives/keeper/perpetual_futures.go:95]: > add a returning value containing possible loss to liquidity provider?

please do it

File: x/derivatives/types/positions.go (hunk 2 above)
Comment (mkXultra) [reply x/derivatives/types/positions.go:86]: ??
Sorry. I was confused.
Please ignore this Msg.

File: x/derivatives/keeper/perpetual_futures.go (hunk 1 above)
Comment (mkXultra) [reply x/derivatives/keeper/perpetual_futures.go:85]: oh, ok.
plz make issuen and write comment.

Comment (mkXultra) [reply x/derivatives/keeper/perpetual_futures.go:85]: This PR does not fix that.

File: x/derivatives/keeper/perpetual_futures.go (hunk 3 above)
Comment (taiki1frsh) [reply x/derivatives/keeper/perpetual_futures.go]: done.
plz check

**Pull request merged**